File: machine/cpp/machine/devices/device_status.cpp
#include "machine/devices/device_status.h"

namespace bmsx {

DeviceStatusLatch::DeviceStatusLatch(Memory& memory, DeviceStatusRegisters registers)
	: code(registers.noneCode)
	, m_memory(memory)
	, m_registers(registers) {
	detail = 0u;
}

void DeviceStatusLatch::resetStatus() const {
	status = 0u;
	code = m_registers.noneCode;
	detail = 0u;
	writeRegisterState();
}

void DeviceStatusLatch::restore(uint32_t status, uint32_t code, uint32_t detail) const {
	this->status = status;
	this->code = code;
	this->detail = detail;
	writeRegisterState();
}

void DeviceStatusLatch::writeRegisterState() const {
	m_memory.writeIoU32(m_registers.statusAddr, status);
	m_memory.writeIoU32(m_registers.codeAddr, code);
	m_memory.writeIoU32(m_registers.detailAddr, detail);
	m_memory.writeIoU32(m_registers.ackAddr, 0u);
}

void DeviceStatusLatch::setStatusFlag(uint32_t mask, bool active) const {
	const uint32_t nextStatus = active ? (status | mask) : (status & ~mask);
	if (nextStatus == status) {
		return;
	}
	status = nextStatus;
	m_memory.writeIoU32(m_registers.statusAddr, status);
}

void DeviceStatusLatch::raise(uint32_t code, uint32_t detail) const {
	if ((status & m_registers.faultMask) != 0u) {
		return;
	}
	this->code = code;
	this->detail = detail;
	m_memory.writeIoU32(m_registers.codeAddr, code);
	m_memory.writeIoU32(m_registers.detailAddr, detail);
	setStatusFlag(m_registers.faultMask, true);
}

void DeviceStatusLatch::acknowledgeWriteThunk(void* context, uint32_t, u32 value, MappedBusSignals) {
	if (value == 0u) {
		return;
	}
	auto* latch = static_cast<DeviceStatusLatch*>(context);
	latch->code = latch->m_registers.noneCode;
	latch->detail = 0u;
	latch->m_memory.writeIoU32(latch->m_registers.codeAddr, latch->code);
	latch->m_memory.writeIoU32(latch->m_registers.detailAddr, latch->detail);
	latch->setStatusFlag(latch->m_registers.faultMask, false);
	latch->m_memory.writeIoU32(latch->m_registers.ackAddr, 0u);
}

} // namespace bmsx

File: machine/ts/machine/devices/device_status.ts
import type { Memory } from '../memory/memory';

export type DeviceStatusRegisters = {
	statusAddr: number;
	codeAddr: number;
	detailAddr: number;
	ackAddr: number;
	faultMask: number;
	noneCode: number;
};

export class DeviceStatusLatch {
	public status = 0;
	public code: number;
	public detail = 0;

	public constructor(
		private readonly memory: Memory,
		private readonly registers: DeviceStatusRegisters,
	) {
		this.code = registers.noneCode;
		this.detail = 0;
	}

	public resetStatus(): void {
		this.status = 0;
		this.code = this.registers.noneCode;
		this.detail = 0;
		this.writeRegisterState();
	}

	public restore(status: number, code: number, detail: number): void {
		this.status = status >>> 0;
		this.code = code >>> 0;
		this.detail = detail >>> 0;
		this.writeRegisterState();
	}

	private writeRegisterState(): void {
		this.memory.writeIoU32(this.registers.statusAddr, this.status);
		this.memory.writeIoU32(this.registers.codeAddr, this.code);
		this.memory.writeIoU32(this.registers.detailAddr, this.detail);
		this.memory.writeIoU32(this.registers.ackAddr, 0);
	}

	public static acknowledgeWriteThunk(context: DeviceStatusLatch, _address: number, value: number): void {
		if (value === 0) {
			return;
		}
		context.code = context.registers.noneCode;
		context.detail = 0;
		context.memory.writeIoU32(context.registers.codeAddr, context.code);
		context.memory.writeIoU32(context.registers.detailAddr, context.detail);
		context.setStatusFlag(context.registers.faultMask, false);
		context.memory.writeIoU32(context.registers.ackAddr, 0);
	}

	public setStatusFlag(mask: number, active: boolean): void {
		const nextStatus = active ? (this.status | mask) : (this.status & ~mask);
		if (nextStatus === this.status) {
			return;
		}
		this.status = nextStatus >>> 0;
		this.memory.writeIoU32(this.registers.statusAddr, this.status);
	}

	public raise(code: number, detail: number): void {
		if ((this.status & this.registers.faultMask) !== 0) {
			return;
		}
		this.code = code >>> 0;
		this.detail = detail >>> 0;
		this.memory.writeIoU32(this.registers.codeAddr, this.code);
		this.memory.writeIoU32(this.registers.detailAddr, this.detail);
		this.setStatusFlag(this.registers.faultMask, true);
	}
}
